src/ui/PipelineContext/PipelineContext.ts

```diff
@@ -44,63 +46,97 @@ export interface IPipelineContextOptions {}
+    this.setContext(
+      $$(this.element)
+        .text()
+        .trim()
+    );
+    this.bind.onRootElement(QueryEvents.buildingQuery, (args: IBuildingQueryEventArgs) => this.handleBuildingQuery(args));
+  }
 
-    if (this.element.tagName.toLowerCase() == 'script') {
-      try {
-        // Content can be HTML encoded for special char ({!})
-        this.content = JSON.parse(Utils.decodeHTMLEntities($$(this.element).text()));
-      } catch (e) {
-        try {
-          this.content = JSON.parse($$(this.element).text());
-        } catch (e) {
-          return;
-        }
-      }
+  /**
+   * Set a new context, replacing any value previously set.
+   * 
+   * @param newContext The new context to set, which can be directly passed as a JSON, or as a stringified JSON.
+   */
+  public setContext(newContext: string | Context) {
+    if (_.isString(newContext)) {
+      const contextParsed = this.tryParseContextFromString(newContext);
+      this.contextContent = contextParsed;
+    } else {
+      this.contextContent = newContext;
     }
-    this.bind.onRootElement(QueryEvents.buildingQuery, (args: IBuildingQueryEventArgs) => this.handleBuildingQuery(args));
+  }
+
+  /**
+   * Return the current context
+   */
+  public getContext(): Context {
+    const keys = this.getContextKeys();
+    return _.object(keys, _.map(keys, key => this.getContextValue(key)));
+  }
+
+  /**
+   * Set a value for a context, replacing the old value if it was already available.
+   * @param contextKey
+   * @param contextValue 
+   */
+  public setContextValue(contextKey: string, contextValue: string | string[]) {
+    this.contextContent[contextKey] = contextValue;
   }
 
   /**
    * Return all the context keys configured for context.
-   * @returns {string[]|Array}
+   * @returns {string[]}
    */
   public getContextKeys(): string[] {
-    return this.content ? _.keys(this.content) : [];
+    return _.keys(this.contextContent);
   }
 
   /**
    * Get the context value associated to the given key.
+   * 
+   * If the global variable Coveo.context contains the requested key, this method will return the value contained in Coveo.context instead of the one contained internally.
+   * 
+   * This is especially useful in a Coveo for Salesforce context, where context values can be extracted from a backend service.
    * @param key
    * @returns {string}
    */
   public getContextValue(key: string): string | string[] {
-    const contextValue = this.content[key];
+    const contextValue = this.contextContent[key];
     if (_.isArray(contextValue)) {
       const contextValues = [];
-      _.each(this.content[key], value => {
+      _.each(this.contextContent[key], value => {
         contextValues.push(this.getModifiedData(value));
       });
       return contextValues;
     } else if (_.isString(contextValue)) {
       return this.getModifiedData(contextValue);
     }
+    return undefined;
```

Comment:
Is there a reason for choosing `undefined` over `null` or an empty string? With `undefined` it is not clear what happened if debugging. e.g. did we forget to execute `getContextValue`? or did it fail mid-execution?